// @aws-cdk/aws-ecs/test/test.splunk-log-driver.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/core");
const ecs = require("../lib");
let stack;
let td;
const image = ecs.ContainerImage.fromRegistry('test-image');
module.exports = {
    'setUp'(cb) {
        stack = new cdk.Stack();
        td = new ecs.Ec2TaskDefinition(stack, 'TaskDefinition');
        cb();
    },
    'create a splunk log driver with minimum options'(test) {
        // WHEN
        td.addContainer('Container', {
            image,
            logging: new ecs.SplunkLogDriver({
                token: cdk.SecretValue.secretsManager('my-splunk-token'),
                url: 'my-splunk-url'
            }),
            memoryLimitMiB: 128
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    LogConfiguration: {
                        LogDriver: 'splunk',
                        Options: {
                            'splunk-token': '{{resolve:secretsmanager:my-splunk-token:SecretString:::}}',
                            'splunk-url': 'my-splunk-url'
                        }
                    }
                }
            ]
        }));
        test.done();
    },
    "create a splunk log driver using splunk with minimum options"(test) {
        // WHEN
        td.addContainer('Container', {
            image,
            logging: ecs.LogDrivers.splunk({
                token: cdk.SecretValue.secretsManager('my-splunk-token'),
                url: 'my-splunk-url'
            }),
            memoryLimitMiB: 128
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    LogConfiguration: {
                        LogDriver: 'splunk',
                        Options: {
                            'splunk-token': '{{resolve:secretsmanager:my-splunk-token:SecretString:::}}',
                            'splunk-url': 'my-splunk-url'
                        }
                    }
                }
            ]
        }));
        test.done();
    },
    "create a splunk log driver using splunk with sourcetype defined"(test) {
        // WHEN
        td.addContainer('Container', {
            image,
            logging: ecs.LogDrivers.splunk({
                token: cdk.SecretValue.secretsManager('my-splunk-token'),
                url: 'my-splunk-url',
                sourceType: 'my-source-type'
            }),
            memoryLimitMiB: 128
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    LogConfiguration: {
                        LogDriver: 'splunk',
                        Options: {
                            'splunk-token': '{{resolve:secretsmanager:my-splunk-token:SecretString:::}}',
                            'splunk-url': 'my-splunk-url',
                            'splunk-sourcetype': 'my-source-type'
                        }
                    }
                }
            ]
        }));
        test.done();
    },
};
