// @aws-cdk/aws-stepfunctions/test/test.state-machine.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/core");
const stepfunctions = require("../lib");
module.exports = {
    'Instantiate Default State Machine'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new stepfunctions.StateMachine(stack, 'MyStateMachine', {
            stateMachineName: "MyStateMachine",
            definition: stepfunctions.Chain.start(new stepfunctions.Pass(stack, 'Pass'))
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::StepFunctions::StateMachine', {
            StateMachineName: "MyStateMachine",
            DefinitionString: "{\"StartAt\":\"Pass\",\"States\":{\"Pass\":{\"Type\":\"Pass\",\"End\":true}}}"
        }));
        test.done();
    },
    'Instantiate Standard State Machine'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new stepfunctions.StateMachine(stack, 'MyStateMachine', {
            stateMachineName: "MyStateMachine",
            definition: stepfunctions.Chain.start(new stepfunctions.Pass(stack, 'Pass')),
            stateMachineType: stepfunctions.StateMachineType.STANDARD
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::StepFunctions::StateMachine', {
            StateMachineName: "MyStateMachine",
            StateMachineType: "STANDARD",
            DefinitionString: "{\"StartAt\":\"Pass\",\"States\":{\"Pass\":{\"Type\":\"Pass\",\"End\":true}}}"
        }));
        test.done();
    },
    'Instantiate Express State Machine'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new stepfunctions.StateMachine(stack, 'MyStateMachine', {
            stateMachineName: "MyStateMachine",
            definition: stepfunctions.Chain.start(new stepfunctions.Pass(stack, 'Pass')),
            stateMachineType: stepfunctions.StateMachineType.EXPRESS
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::StepFunctions::StateMachine', {
            StateMachineName: "MyStateMachine",
            StateMachineType: "EXPRESS",
            DefinitionString: "{\"StartAt\":\"Pass\",\"States\":{\"Pass\":{\"Type\":\"Pass\",\"End\":true}}}"
        }));
        test.done();
    }
};
