// @aws-cdk/aws-eks/test/test.k8s-patch.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const core_1 = require("@aws-cdk/core");
const eks = require("../lib");
const k8s_patch_1 = require("../lib/k8s-patch");
module.exports = {
    'applies a patch to k8s'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        const cluster = new eks.Cluster(stack, 'MyCluster');
        // WHEN
        new k8s_patch_1.KubernetesPatch(stack, 'MyPatch', {
            cluster,
            applyPatch: { patch: { to: 'apply' } },
            restorePatch: { restore: { patch: 123 } },
            resourceName: 'myResourceName',
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('Custom::AWSCDK-EKS-KubernetesPatch', {
            ServiceToken: {
                "Fn::GetAtt": [
                    "awscdkawseksKubectlProviderNestedStackawscdkawseksKubectlProviderNestedStackResourceA7AEBA6B",
                    "Outputs.awscdkawseksKubectlProviderframeworkonEvent0A650005Arn"
                ]
            },
            ResourceName: "myResourceName",
            ResourceNamespace: "default",
            ApplyPatchJson: "{\"patch\":{\"to\":\"apply\"}}",
            RestorePatchJson: "{\"restore\":{\"patch\":123}}",
            ClusterName: {
                Ref: "MyCluster8AD82BF8"
            },
            RoleArn: {
                "Fn::GetAtt": [
                    "MyClusterCreationRoleB5FA4FF3",
                    "Arn"
                ]
            }
        }));
        test.done();
    }
};
