// @aws-cdk/aws-eks/test/test.cluster.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const ec2 = require("@aws-cdk/aws-ec2");
const iam = require("@aws-cdk/aws-iam");
const cdk = require("@aws-cdk/core");
const eks = require("../lib");
const spot_interrupt_handler_1 = require("../lib/spot-interrupt-handler");
const util_1 = require("./util");
module.exports = {
    'a default cluster spans all subnets'(test) {
        // GIVEN
        const { stack, vpc } = util_1.testFixture();
        // WHEN
        new eks.Cluster(stack, 'Cluster', { vpc, kubectlEnabled: false, defaultCapacity: 0 });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::EKS::Cluster', {
            ResourcesVpcConfig: {
                SubnetIds: [
                    { Ref: "VPCPublicSubnet1SubnetB4246D30" },
                    { Ref: "VPCPublicSubnet2Subnet74179F39" },
                    { Ref: "VPCPrivateSubnet1Subnet8BCA10E0" },
                    { Ref: "VPCPrivateSubnet2SubnetCFCDAA7A" },
                ]
            }
        }));
        test.done();
    },
    'if "vpc" is not specified, vpc with default configuration will be created'(test) {
        // GIVEN
        const { stack } = util_1.testFixtureNoVpc();
        // WHEN
        new eks.Cluster(stack, 'cluster');
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::EC2::VPC'));
        test.done();
    },
    'default capacity': {
        'x2 m5.large by default'(test) {
            // GIVEN
            const { stack } = util_1.testFixtureNoVpc();
            // WHEN
            const cluster = new eks.Cluster(stack, 'cluster');
            // THEN
            test.ok(cluster.defaultCapacity);
            assert_1.expect(stack).to(assert_1.haveResource('AWS::AutoScaling::AutoScalingGroup', { MinSize: '2', MaxSize: '2' }));
            assert_1.expect(stack).to(assert_1.haveResource('AWS::AutoScaling::LaunchConfiguration', { InstanceType: 'm5.large' }));
            test.done();
        },
        'quantity and type can be customized'(test) {
            // GIVEN
            const { stack } = util_1.testFixtureNoVpc();
            // WHEN
            const cluster = new eks.Cluster(stack, 'cluster', {
                defaultCapacity: 10,
                defaultCapacityInstance: new ec2.InstanceType('m2.xlarge')
            });
            // THEN
            test.ok(cluster.defaultCapacity);
            assert_1.expect(stack).to(assert_1.haveResource('AWS::AutoScaling::AutoScalingGroup', { MinSize: '10', MaxSize: '10' }));
            assert_1.expect(stack).to(assert_1.haveResource('AWS::AutoScaling::LaunchConfiguration', { InstanceType: 'm2.xlarge' }));
            test.done();
        },
        'defaultCapacity=0 will not allocate at all'(test) {
            // GIVEN
            const { stack } = util_1.testFixtureNoVpc();
            // WHEN
            const cluster = new eks.Cluster(stack, 'cluster', { defaultCapacity: 0 });
            // THEN
            test.ok(!cluster.defaultCapacity);
            assert_1.expect(stack).notTo(assert_1.haveResource('AWS::AutoScaling::AutoScalingGroup'));
            assert_1.expect(stack).notTo(assert_1.haveResource('AWS::AutoScaling::LaunchConfiguration'));
            test.done();
        }
    },
    'creating a cluster tags the private VPC subnets'(test) {
        // GIVEN
        const { stack, vpc } = util_1.testFixture();
        // WHEN
        new eks.Cluster(stack, 'Cluster', { vpc, kubectlEnabled: false, defaultCapacity: 0 });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::EC2::Subnet', {
            Tags: [
                { Key: "Name", Value: "Stack/VPC/PrivateSubnet1" },
                { Key: "aws-cdk:subnet-name", Value: "Private" },
                { Key: "aws-cdk:subnet-type", Value: "Private" },
                { Key: "kubernetes.io/role/internal-elb", Value: "1" }
            ]
        }));
        test.done();
    },
    'creating a cluster tags the public VPC subnets'(test) {
        // GIVEN
        const { stack, vpc } = util_1.testFixture();
        // WHEN
        new eks.Cluster(stack, 'Cluster', { vpc, kubectlEnabled: false, defaultCapacity: 0 });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::EC2::Subnet', {
            MapPublicIpOnLaunch: true,
            Tags: [
                { Key: "Name", Value: "Stack/VPC/PublicSubnet1" },
                { Key: "aws-cdk:subnet-name", Value: "Public" },
                { Key: "aws-cdk:subnet-type", Value: "Public" },
                { Key: "kubernetes.io/role/elb", Value: "1" }
            ]
        }));
        test.done();
    },
    'adding capacity creates an ASG with tags'(test) {
        // GIVEN
        const { stack, vpc } = util_1.testFixture();
        const cluster = new eks.Cluster(stack, 'Cluster', { vpc, kubectlEnabled: false, defaultCapacity: 0 });
        // WHEN
        cluster.addCapacity('Default', {
            instanceType: new ec2.InstanceType('t2.medium'),
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::AutoScaling::AutoScalingGroup', {
            Tags: [
                {
                    Key: "Name",
                    PropagateAtLaunch: true,
                    Value: "Stack/Cluster/Default"
                },
                {
                    Key: { "Fn::Join": ["", ["kubernetes.io/cluster/", { Ref: "ClusterEB0386A7" }]] },
                    PropagateAtLaunch: true,
                    Value: "owned"
                }
            ]
        }));
        test.done();
    },
    'exercise export/import'(test) {
        // GIVEN
        const { stack: stack1, vpc, app } = util_1.testFixture();
        const stack2 = new cdk.Stack(app, 'stack2', { env: { region: 'us-east-1' } });
        const cluster = new eks.Cluster(stack1, 'Cluster', { vpc, kubectlEnabled: false, defaultCapacity: 0 });
        // WHEN
        const imported = eks.Cluster.fromClusterAttributes(stack2, 'Imported', {
            clusterArn: cluster.clusterArn,
            vpc: cluster.vpc,
            clusterEndpoint: cluster.clusterEndpoint,
            clusterName: cluster.clusterName,
            securityGroups: cluster.connections.securityGroups,
            clusterCertificateAuthorityData: cluster.clusterCertificateAuthorityData
        });
        // this should cause an export/import
        new cdk.CfnOutput(stack2, 'ClusterARN', { value: imported.clusterArn });
        // THEN
        assert_1.expect(stack2).toMatch({
            Outputs: {
                ClusterARN: {
                    Value: {
                        "Fn::ImportValue": "Stack:ExportsOutputFnGetAttClusterEB0386A7Arn2F2E3C3F"
                    }
                }
            }
        });
        test.done();
    },
    'disabled features when kubectl is disabled'(test) {
        // GIVEN
        const { stack, vpc } = util_1.testFixture();
        const cluster = new eks.Cluster(stack, 'Cluster', { vpc, kubectlEnabled: false, defaultCapacity: 0 });
        test.throws(() => cluster.awsAuth, /Cannot define aws-auth mappings if kubectl is disabled/);
        test.throws(() => cluster.addResource('foo', {}), /Unable to perform this operation since kubectl is not enabled for this cluster/);
        test.throws(() => cluster.addCapacity('boo', { instanceType: new ec2.InstanceType('r5d.24xlarge'), mapRole: true }), /Cannot map instance IAM role to RBAC if kubectl is disabled for the cluster/);
        test.throws(() => new eks.HelmChart(stack, 'MyChart', { cluster, chart: 'chart' }), /Unable to perform this operation since kubectl is not enabled for this cluster/);
        test.done();
    },
    'mastersRole can be used to map an IAM role to "system:masters" (required kubectl)'(test) {
        // GIVEN
        const { stack, vpc } = util_1.testFixture();
        const role = new iam.Role(stack, 'role', { assumedBy: new iam.AnyPrincipal() });
        // WHEN
        new eks.Cluster(stack, 'Cluster', { vpc, mastersRole: role, defaultCapacity: 0 });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource(eks.KubernetesResource.RESOURCE_TYPE, {
            Manifest: {
                "Fn::Join": [
                    "",
                    [
                        "[{\"apiVersion\":\"v1\",\"kind\":\"ConfigMap\",\"metadata\":{\"name\":\"aws-auth\",\"namespace\":\"kube-system\"},\"data\":{\"mapRoles\":\"[{\\\"rolearn\\\":\\\"",
                        {
                            "Fn::GetAtt": [
                                "roleC7B7E775",
                                "Arn"
                            ]
                        },
                        "\\\",\\\"username\\\":\\\"",
                        {
                            "Fn::GetAtt": [
                                "roleC7B7E775",
                                "Arn"
                            ]
                        },
                        "\\\",\\\"groups\\\":[\\\"system:masters\\\"]}]\",\"mapUsers\":\"[]\",\"mapAccounts\":\"[]\"}}]"
                    ]
                ]
            }
        }));
        test.done();
    },
    'addResource can be used to apply k8s manifests on this cluster'(test) {
        // GIVEN
        const { stack, vpc } = util_1.testFixture();
        const cluster = new eks.Cluster(stack, 'Cluster', { vpc, defaultCapacity: 0 });
        // WHEN
        cluster.addResource('manifest1', { foo: 123 });
        cluster.addResource('manifest2', { bar: 123 }, { boor: [1, 2, 3] });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource(eks.KubernetesResource.RESOURCE_TYPE, {
            Manifest: "[{\"foo\":123}]"
        }));
        assert_1.expect(stack).to(assert_1.haveResource(eks.KubernetesResource.RESOURCE_TYPE, {
            Manifest: "[{\"bar\":123},{\"boor\":[1,2,3]}]"
        }));
        test.done();
    },
    'when kubectl is enabled (default) adding capacity will automatically map its IAM role'(test) {
        // GIVEN
        const { stack, vpc } = util_1.testFixture();
        const cluster = new eks.Cluster(stack, 'Cluster', { vpc, defaultCapacity: 0 });
        // WHEN
        cluster.addCapacity('default', {
            instanceType: new ec2.InstanceType('t2.nano'),
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource(eks.KubernetesResource.RESOURCE_TYPE, {
            Manifest: {
                "Fn::Join": [
                    "",
                    [
                        "[{\"apiVersion\":\"v1\",\"kind\":\"ConfigMap\",\"metadata\":{\"name\":\"aws-auth\",\"namespace\":\"kube-system\"},\"data\":{\"mapRoles\":\"[{\\\"rolearn\\\":\\\"",
                        {
                            "Fn::GetAtt": [
                                "ClusterdefaultInstanceRoleF20A29CD",
                                "Arn"
                            ]
                        },
                        "\\\",\\\"username\\\":\\\"system:node:{{EC2PrivateDNSName}}\\\",\\\"groups\\\":[\\\"system:bootstrappers\\\",\\\"system:nodes\\\"]}]\",\"mapUsers\":\"[]\",\"mapAccounts\":\"[]\"}}]"
                    ]
                ]
            }
        }));
        test.done();
    },
    'addCapacity will *not* map the IAM role if mapRole is false'(test) {
        // GIVEN
        const { stack, vpc } = util_1.testFixture();
        const cluster = new eks.Cluster(stack, 'Cluster', { vpc, defaultCapacity: 0 });
        // WHEN
        cluster.addCapacity('default', {
            instanceType: new ec2.InstanceType('t2.nano'),
            mapRole: false
        });
        // THEN
        assert_1.expect(stack).to(assert_1.not(assert_1.haveResource(eks.KubernetesResource.RESOURCE_TYPE)));
        test.done();
    },
    'addCapacity will *not* map the IAM role if kubectl is disabled'(test) {
        // GIVEN
        const { stack, vpc } = util_1.testFixture();
        const cluster = new eks.Cluster(stack, 'Cluster', { vpc, kubectlEnabled: false, defaultCapacity: 0 });
        // WHEN
        cluster.addCapacity('default', {
            instanceType: new ec2.InstanceType('t2.nano')
        });
        // THEN
        assert_1.expect(stack).to(assert_1.not(assert_1.haveResource(eks.KubernetesResource.RESOURCE_TYPE)));
        test.done();
    },
    'outputs': {
        'aws eks update-kubeconfig is the only output synthesized by default'(test) {
            // GIVEN
            const { app, stack } = util_1.testFixtureNoVpc();
            // WHEN
            new eks.Cluster(stack, 'Cluster');
            // THEN
            const assembly = app.synth();
            const template = assembly.getStackByName(stack.stackName).template;
            test.deepEqual(template.Outputs, {
                ClusterConfigCommand43AAE40F: { Value: { 'Fn::Join': ['', ['aws eks update-kubeconfig --name ', { Ref: 'Cluster9EE0221C' }, ' --region us-east-1']] } },
                ClusterGetTokenCommand06AE992E: { Value: { 'Fn::Join': ['', ['aws eks get-token --cluster-name ', { Ref: 'Cluster9EE0221C' }, ' --region us-east-1']] } }
            });
            test.done();
        },
        'if masters role is defined, it should be included in the config command'(test) {
            // GIVEN
            const { app, stack } = util_1.testFixtureNoVpc();
            // WHEN
            const mastersRole = new iam.Role(stack, 'masters', { assumedBy: new iam.AccountRootPrincipal() });
            new eks.Cluster(stack, 'Cluster', { mastersRole });
            // THEN
            const assembly = app.synth();
            const template = assembly.getStackByName(stack.stackName).template;
            test.deepEqual(template.Outputs, {
                ClusterConfigCommand43AAE40F: { Value: { 'Fn::Join': ['', ['aws eks update-kubeconfig --name ', { Ref: 'Cluster9EE0221C' }, ' --region us-east-1 --role-arn ', { 'Fn::GetAtt': ['masters0D04F23D', 'Arn'] }]] } },
                ClusterGetTokenCommand06AE992E: { Value: { 'Fn::Join': ['', ['aws eks get-token --cluster-name ', { Ref: 'Cluster9EE0221C' }, ' --region us-east-1 --role-arn ', { 'Fn::GetAtt': ['masters0D04F23D', 'Arn'] }]] } }
            });
            test.done();
        },
        'if `outputConfigCommand=false` will disabled the output'(test) {
            // GIVEN
            const { app, stack } = util_1.testFixtureNoVpc();
            // WHEN
            const mastersRole = new iam.Role(stack, 'masters', { assumedBy: new iam.AccountRootPrincipal() });
            new eks.Cluster(stack, 'Cluster', {
                mastersRole,
                outputConfigCommand: false,
            });
            // THEN
            const assembly = app.synth();
            const template = assembly.getStackByName(stack.stackName).template;
            test.ok(!template.Outputs); // no outputs
            test.done();
        },
        '`outputClusterName` can be used to synthesize an output with the cluster name'(test) {
            // GIVEN
            const { app, stack } = util_1.testFixtureNoVpc();
            // WHEN
            new eks.Cluster(stack, 'Cluster', {
                outputConfigCommand: false,
                outputClusterName: true
            });
            // THEN
            const assembly = app.synth();
            const template = assembly.getStackByName(stack.stackName).template;
            test.deepEqual(template.Outputs, {
                ClusterClusterNameEB26049E: { Value: { Ref: 'Cluster9EE0221C' } }
            });
            test.done();
        },
        '`outputMastersRoleArn` can be used to synthesize an output with the arn of the masters role if defined'(test) {
            // GIVEN
            const { app, stack } = util_1.testFixtureNoVpc();
            // WHEN
            new eks.Cluster(stack, 'Cluster', {
                outputConfigCommand: false,
                outputMastersRoleArn: true,
                mastersRole: new iam.Role(stack, 'masters', { assumedBy: new iam.AccountRootPrincipal() })
            });
            // THEN
            const assembly = app.synth();
            const template = assembly.getStackByName(stack.stackName).template;
            test.deepEqual(template.Outputs, {
                ClusterMastersRoleArnB15964B1: { Value: { 'Fn::GetAtt': ['masters0D04F23D', 'Arn'] } }
            });
            test.done();
        },
        'when adding capacity, instance role ARN will not be outputed only if we do not auto-map aws-auth'(test) {
            // GIVEN
            const { app, stack } = util_1.testFixtureNoVpc();
            // WHEN
            new eks.Cluster(stack, 'Cluster', {
                outputConfigCommand: false,
                kubectlEnabled: false
            });
            // THEN
            const assembly = app.synth();
            const template = assembly.getStackByName(stack.stackName).template;
            test.deepEqual(template.Outputs, {
                ClusterDefaultCapacityInstanceRoleARN7DADF219: {
                    Value: { 'Fn::GetAtt': ['ClusterDefaultCapacityInstanceRole3E209969', 'Arn'] }
                }
            });
            test.done();
        },
    },
    'boostrap user-data': {
        'rendered by default for ASGs'(test) {
            // GIVEN
            const { app, stack } = util_1.testFixtureNoVpc();
            const cluster = new eks.Cluster(stack, 'Cluster', { defaultCapacity: 0 });
            // WHEN
            cluster.addCapacity('MyCapcity', { instanceType: new ec2.InstanceType('m3.xlargs') });
            // THEN
            const template = app.synth().getStackByName(stack.stackName).template;
            const userData = template.Resources.ClusterMyCapcityLaunchConfig58583345.Properties.UserData;
            test.deepEqual(userData, { 'Fn::Base64': { 'Fn::Join': ['', ['#!/bin/bash\nset -o xtrace\n/etc/eks/bootstrap.sh ', { Ref: 'Cluster9EE0221C' }, ' --kubelet-extra-args "--node-labels lifecycle=OnDemand" --use-max-pods true\n/opt/aws/bin/cfn-signal --exit-code $? --stack Stack --resource ClusterMyCapcityASGD4CD8B97 --region us-east-1']] } });
            test.done();
        },
        'not rendered if bootstrap is disabled'(test) {
            // GIVEN
            const { app, stack } = util_1.testFixtureNoVpc();
            const cluster = new eks.Cluster(stack, 'Cluster', { defaultCapacity: 0 });
            // WHEN
            cluster.addCapacity('MyCapcity', {
                instanceType: new ec2.InstanceType('m3.xlargs'),
                bootstrapEnabled: false
            });
            // THEN
            const template = app.synth().getStackByName(stack.stackName).template;
            const userData = template.Resources.ClusterMyCapcityLaunchConfig58583345.Properties.UserData;
            test.deepEqual(userData, { "Fn::Base64": "#!/bin/bash" });
            test.done();
        },
        // cursory test for options: see test.user-data.ts for full suite
        'bootstrap options'(test) {
            // GIVEN
            const { app, stack } = util_1.testFixtureNoVpc();
            const cluster = new eks.Cluster(stack, 'Cluster', { defaultCapacity: 0 });
            // WHEN
            cluster.addCapacity('MyCapcity', {
                instanceType: new ec2.InstanceType('m3.xlargs'),
                bootstrapOptions: {
                    kubeletExtraArgs: '--node-labels FOO=42'
                }
            });
            // THEN
            const template = app.synth().getStackByName(stack.stackName).template;
            const userData = template.Resources.ClusterMyCapcityLaunchConfig58583345.Properties.UserData;
            test.deepEqual(userData, { 'Fn::Base64': { 'Fn::Join': ['', ['#!/bin/bash\nset -o xtrace\n/etc/eks/bootstrap.sh ', { Ref: 'Cluster9EE0221C' }, ' --kubelet-extra-args "--node-labels lifecycle=OnDemand  --node-labels FOO=42" --use-max-pods true\n/opt/aws/bin/cfn-signal --exit-code $? --stack Stack --resource ClusterMyCapcityASGD4CD8B97 --region us-east-1']] } });
            test.done();
        },
        'spot instances': {
            'nodes labeled an tainted accordingly'(test) {
                // GIVEN
                const { app, stack } = util_1.testFixtureNoVpc();
                const cluster = new eks.Cluster(stack, 'Cluster', { defaultCapacity: 0 });
                // WHEN
                cluster.addCapacity('MyCapcity', {
                    instanceType: new ec2.InstanceType('m3.xlargs'),
                    spotPrice: '0.01'
                });
                // THEN
                const template = app.synth().getStackByName(stack.stackName).template;
                const userData = template.Resources.ClusterMyCapcityLaunchConfig58583345.Properties.UserData;
                test.deepEqual(userData, { 'Fn::Base64': { 'Fn::Join': ['', ['#!/bin/bash\nset -o xtrace\n/etc/eks/bootstrap.sh ', { Ref: 'Cluster9EE0221C' }, ' --kubelet-extra-args "--node-labels lifecycle=Ec2Spot --register-with-taints=spotInstance=true:PreferNoSchedule" --use-max-pods true\n/opt/aws/bin/cfn-signal --exit-code $? --stack Stack --resource ClusterMyCapcityASGD4CD8B97 --region us-east-1']] } });
                test.done();
            },
            'if kubectl is enabled, the interrupt handler is added'(test) {
                // GIVEN
                const { stack } = util_1.testFixtureNoVpc();
                const cluster = new eks.Cluster(stack, 'Cluster', { defaultCapacity: 0 });
                // WHEN
                cluster.addCapacity('MyCapcity', {
                    instanceType: new ec2.InstanceType('m3.xlargs'),
                    spotPrice: '0.01'
                });
                // THEN
                assert_1.expect(stack).to(assert_1.haveResource(eks.KubernetesResource.RESOURCE_TYPE, { Manifest: JSON.stringify(spot_interrupt_handler_1.spotInterruptHandler()) }));
                test.done();
            },
            'if kubectl is disabled, interrupt handler is not added'(test) {
                // GIVEN
                const { stack } = util_1.testFixtureNoVpc();
                const cluster = new eks.Cluster(stack, 'Cluster', { defaultCapacity: 0, kubectlEnabled: false });
                // WHEN
                cluster.addCapacity('MyCapcity', {
                    instanceType: new ec2.InstanceType('m3.xlargs'),
                    spotPrice: '0.01'
                });
                // THEN
                assert_1.expect(stack).notTo(assert_1.haveResource(eks.KubernetesResource.RESOURCE_TYPE));
                test.done();
            }
        }
    },
    'if bootstrap is disabled cannot specify options'(test) {
        // GIVEN
        const { stack } = util_1.testFixtureNoVpc();
        const cluster = new eks.Cluster(stack, 'Cluster', { defaultCapacity: 0 });
        // THEN
        test.throws(() => cluster.addCapacity('MyCapcity', {
            instanceType: new ec2.InstanceType('m3.xlargs'),
            bootstrapEnabled: false,
            bootstrapOptions: { awsApiRetryAttempts: 10 }
        }), /Cannot specify "bootstrapOptions" if "bootstrapEnabled" is false/);
        test.done();
    },
    'EKS-Optimized AMI with GPU support'(test) {
        // GIVEN
        const { app, stack } = util_1.testFixtureNoVpc();
        // WHEN
        new eks.Cluster(stack, 'cluster', {
            defaultCapacity: 2,
            defaultCapacityInstance: new ec2.InstanceType('g4dn.xlarge'),
        });
        // THEN
        const assembly = app.synth();
        const parameters = assembly.getStackByName(stack.stackName).template.Parameters;
        test.ok(Object.entries(parameters).some(([k, v]) => k.startsWith('SsmParameterValueawsserviceeksoptimizedami') && v.Default.includes('amazon-linux2-gpu')), 'EKS AMI with GPU should be in ssm parameters');
        test.done();
    },
    'when using custom resource a creation role & policy is defined'(test) {
        // GIVEN
        const { stack } = util_1.testFixture();
        // WHEN
        new eks.Cluster(stack, 'MyCluster', {
            clusterName: 'my-cluster-name'
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('Custom::AWSCDK-EKS-Cluster', {
            Config: {
                name: "my-cluster-name",
                roleArn: { "Fn::GetAtt": ["MyClusterRoleBA20FE72", "Arn"] },
                resourcesVpcConfig: {
                    securityGroupIds: [{ "Fn::GetAtt": ["MyClusterControlPlaneSecurityGroup6B658F79", "GroupId"] }],
                    subnetIds: [
                        { Ref: "MyClusterDefaultVpcPublicSubnet1SubnetFAE5A9B6" },
                        { Ref: "MyClusterDefaultVpcPublicSubnet2SubnetF6D028A0" },
                        { Ref: "MyClusterDefaultVpcPrivateSubnet1SubnetE1D0DCDB" },
                        { Ref: "MyClusterDefaultVpcPrivateSubnet2Subnet11FEA8D0" }
                    ]
                }
            }
        }));
        // role can be assumed by 3 lambda handlers (2 for the cluster resource and 1 for the kubernetes resource)
        assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Role', {
            AssumeRolePolicyDocument: {
                Statement: [
                    {
                        Action: "sts:AssumeRole",
                        Effect: "Allow",
                        Principal: {
                            AWS: [
                                {
                                    "Fn::GetAtt": [
                                        "awscdkawseksClusterResourceProviderNestedStackawscdkawseksClusterResourceProviderNestedStackResource9827C454",
                                        "Outputs.StackawscdkawseksClusterResourceProviderOnEventHandlerServiceRole3AEE0A43Arn"
                                    ]
                                },
                                {
                                    "Fn::GetAtt": [
                                        "awscdkawseksClusterResourceProviderNestedStackawscdkawseksClusterResourceProviderNestedStackResource9827C454",
                                        "Outputs.StackawscdkawseksClusterResourceProviderIsCompleteHandlerServiceRole8E7F1C11Arn"
                                    ]
                                }
                            ]
                        }
                    },
                    {
                        Action: "sts:AssumeRole",
                        Effect: "Allow",
                        Principal: {
                            AWS: {
                                "Fn::GetAtt": [
                                    "awscdkawseksKubectlProviderNestedStackawscdkawseksKubectlProviderNestedStackResourceA7AEBA6B",
                                    "Outputs.StackawscdkawseksKubectlProviderHandlerServiceRole2C52B3ECArn"
                                ]
                            }
                        }
                    }
                ],
                Version: "2012-10-17"
            }
        }));
        // policy allows creation role to pass the cluster role and to interact with the cluster (given we know the explicit cluster name)
        assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Policy', {
            PolicyDocument: {
                Statement: [
                    {
                        Action: "iam:PassRole",
                        Effect: "Allow",
                        Resource: {
                            "Fn::GetAtt": [
                                "MyClusterRoleBA20FE72",
                                "Arn"
                            ]
                        }
                    },
                    {
                        Action: "ec2:DescribeSubnets",
                        Effect: "Allow",
                        Resource: "*",
                    },
                    {
                        Action: [
                            "eks:CreateCluster",
                            "eks:DescribeCluster",
                            "eks:DeleteCluster",
                            "eks:UpdateClusterVersion",
                            "eks:UpdateClusterConfig",
                            "eks:CreateFargateProfile"
                        ],
                        Effect: "Allow",
                        Resource: {
                            "Fn::Join": [
                                "",
                                [
                                    "arn:",
                                    {
                                        Ref: "AWS::Partition"
                                    },
                                    ":eks:us-east-1:",
                                    {
                                        Ref: "AWS::AccountId"
                                    },
                                    ":cluster/my-cluster-name"
                                ]
                            ]
                        }
                    },
                    {
                        Action: [
                            "eks:DescribeFargateProfile",
                            "eks:DeleteFargateProfile"
                        ],
                        Effect: "Allow",
                        Resource: {
                            "Fn::Join": [
                                "",
                                [
                                    "arn:",
                                    {
                                        Ref: "AWS::Partition"
                                    },
                                    ":eks:us-east-1:",
                                    {
                                        Ref: "AWS::AccountId"
                                    },
                                    ":fargateprofile/my-cluster-name/*"
                                ]
                            ]
                        }
                    },
                    {
                        Action: "iam:GetRole",
                        Effect: "Allow",
                        Resource: "*"
                    },
                    {
                        Action: "iam:CreateServiceLinkedRole",
                        Effect: "Allow",
                        Resource: "*"
                    }
                ],
                Version: "2012-10-17"
            }
        }));
        test.done();
    },
    'if an explicit cluster name is not provided, the creation role policy is wider (allows interacting with all clusters)'(test) {
        // GIVEN
        const { stack } = util_1.testFixture();
        // WHEN
        new eks.Cluster(stack, 'MyCluster');
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Policy', {
            PolicyDocument: {
                Statement: [
                    {
                        Action: "iam:PassRole",
                        Effect: "Allow",
                        Resource: {
                            "Fn::GetAtt": [
                                "MyClusterRoleBA20FE72",
                                "Arn"
                            ]
                        }
                    },
                    {
                        Action: "ec2:DescribeSubnets",
                        Effect: "Allow",
                        Resource: "*",
                    },
                    {
                        Action: [
                            "eks:CreateCluster",
                            "eks:DescribeCluster",
                            "eks:DeleteCluster",
                            "eks:UpdateClusterVersion",
                            "eks:UpdateClusterConfig",
                            "eks:CreateFargateProfile"
                        ],
                        Effect: "Allow",
                        Resource: "*"
                    },
                    {
                        Action: [
                            "eks:DescribeFargateProfile",
                            "eks:DeleteFargateProfile"
                        ],
                        Effect: "Allow",
                        Resource: "*"
                    },
                    {
                        Action: "iam:GetRole",
                        Effect: "Allow",
                        Resource: "*"
                    },
                    {
                        Action: "iam:CreateServiceLinkedRole",
                        Effect: "Allow",
                        Resource: "*"
                    }
                ],
                Version: "2012-10-17"
            }
        }));
        test.done();
    },
    'if helm charts are used, its resource provider is allowed to assume the creation role'(test) {
        // GIVEN
        const { stack } = util_1.testFixture();
        const cluster = new eks.Cluster(stack, 'MyCluster', {
            clusterName: 'my-cluster-name'
        });
        // WHEN
        cluster.addChart('MyChart', {
            chart: 'foo'
        });
        // THEN
        // role can be assumed by 4 principals: two for the cluster resource, one
        // for kubernetes resource and one for the helm resource.
        assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Role', {
            AssumeRolePolicyDocument: {
                Statement: [
                    {
                        Action: "sts:AssumeRole",
                        Effect: "Allow",
                        Principal: {
                            AWS: [
                                {
                                    "Fn::GetAtt": [
                                        "awscdkawseksClusterResourceProviderNestedStackawscdkawseksClusterResourceProviderNestedStackResource9827C454",
                                        "Outputs.StackawscdkawseksClusterResourceProviderOnEventHandlerServiceRole3AEE0A43Arn"
                                    ]
                                },
                                {
                                    "Fn::GetAtt": [
                                        "awscdkawseksClusterResourceProviderNestedStackawscdkawseksClusterResourceProviderNestedStackResource9827C454",
                                        "Outputs.StackawscdkawseksClusterResourceProviderIsCompleteHandlerServiceRole8E7F1C11Arn"
                                    ]
                                }
                            ]
                        }
                    },
                    {
                        Action: "sts:AssumeRole",
                        Effect: "Allow",
                        Principal: {
                            AWS: {
                                "Fn::GetAtt": [
                                    "awscdkawseksKubectlProviderNestedStackawscdkawseksKubectlProviderNestedStackResourceA7AEBA6B",
                                    "Outputs.StackawscdkawseksKubectlProviderHandlerServiceRole2C52B3ECArn"
                                ]
                            }
                        }
                    }
                ],
                Version: "2012-10-17"
            }
        }));
        test.done();
    },
    'coreDnsComputeType will patch the coreDNS configuration to use a "fargate" compute type and restore to "ec2" upon removal'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new eks.Cluster(stack, 'MyCluster', {
            coreDnsComputeType: eks.CoreDnsComputeType.FARGATE
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('Custom::AWSCDK-EKS-KubernetesPatch', {
            ResourceName: "deployment/coredns",
            ResourceNamespace: "kube-system",
            ApplyPatchJson: "{\"spec\":{\"template\":{\"metadata\":{\"annotations\":{\"eks.amazonaws.com/compute-type\":\"fargate\"}}}}}",
            RestorePatchJson: "{\"spec\":{\"template\":{\"metadata\":{\"annotations\":{\"eks.amazonaws.com/compute-type\":\"ec2\"}}}}}",
            ClusterName: {
                Ref: "MyCluster8AD82BF8"
            },
            RoleArn: {
                "Fn::GetAtt": [
                    "MyClusterCreationRoleB5FA4FF3",
                    "Arn"
                ]
            }
        }));
        test.done();
    }
};
