// @aws-cdk/aws-stepfunctions-tasks/test/emr-set-cluster-termination-protection.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sfn = require("@aws-cdk/aws-stepfunctions");
const cdk = require("@aws-cdk/core");
const tasks = require("../lib");
let stack;
beforeEach(() => {
    // GIVEN
    stack = new cdk.Stack();
});
test('Set termination protection with static ClusterId and TerminationProtected', () => {
    // WHEN
    const task = new sfn.Task(stack, 'Task', {
        task: new tasks.EmrSetClusterTerminationProtection({
            clusterId: 'ClusterId',
            terminationProtected: false
        })
    });
    // THEN
    expect(stack.resolve(task.toStateJson())).toEqual({
        Type: 'Task',
        Resource: {
            'Fn::Join': [
                '',
                [
                    'arn:',
                    {
                        Ref: 'AWS::Partition',
                    },
                    ':states:::elasticmapreduce:setClusterTerminationProtection',
                ],
            ],
        },
        End: true,
        Parameters: {
            ClusterId: 'ClusterId',
            TerminationProtected: false
        },
    });
});
test('Set termination protection with static ClusterId and TerminationProtected from payload', () => {
    // WHEN
    const task = new sfn.Task(stack, 'Task', {
        task: new tasks.EmrSetClusterTerminationProtection({
            clusterId: 'ClusterId',
            terminationProtected: sfn.TaskInput.fromDataAt('$.TerminationProtected').value
        })
    });
    // THEN
    expect(stack.resolve(task.toStateJson())).toEqual({
        Type: 'Task',
        Resource: {
            'Fn::Join': [
                '',
                [
                    'arn:',
                    {
                        Ref: 'AWS::Partition',
                    },
                    ':states:::elasticmapreduce:setClusterTerminationProtection',
                ],
            ],
        },
        End: true,
        Parameters: {
            'ClusterId': 'ClusterId',
            'TerminationProtected.$': '$.TerminationProtected'
        },
    });
});
test('Set termination protection with ClusterId from payload and static TerminationProtected', () => {
    // WHEN
    const task = new sfn.Task(stack, 'Task', {
        task: new tasks.EmrSetClusterTerminationProtection({
            clusterId: sfn.TaskInput.fromDataAt('$.ClusterId').value,
            terminationProtected: false
        })
    });
    // THEN
    expect(stack.resolve(task.toStateJson())).toEqual({
        Type: 'Task',
        Resource: {
            'Fn::Join': [
                '',
                [
                    'arn:',
                    {
                        Ref: 'AWS::Partition',
                    },
                    ':states:::elasticmapreduce:setClusterTerminationProtection',
                ],
            ],
        },
        End: true,
        Parameters: {
            'ClusterId.$': '$.ClusterId',
            'TerminationProtected': false
        },
    });
});
