// @aws-cdk/aws-stepfunctions-tasks/test/integ.sagemaker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const aws_kms_1 = require("@aws-cdk/aws-kms");
const aws_s3_1 = require("@aws-cdk/aws-s3");
const aws_stepfunctions_1 = require("@aws-cdk/aws-stepfunctions");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
const app = new core_1.App();
const stack = new core_1.Stack(app, 'integ-stepfunctions-sagemaker');
const encryptionKey = new aws_kms_1.Key(stack, 'EncryptionKey', {
    removalPolicy: core_1.RemovalPolicy.DESTROY,
});
const trainingData = new aws_s3_1.Bucket(stack, 'TrainingData', {
    encryption: aws_s3_1.BucketEncryption.KMS,
    encryptionKey,
    removalPolicy: core_1.RemovalPolicy.DESTROY,
});
new aws_stepfunctions_1.StateMachine(stack, 'StateMachine', {
    definition: new aws_stepfunctions_1.Task(stack, 'TrainTask', {
        task: new lib_1.SagemakerTrainTask({
            algorithmSpecification: {
                algorithmName: 'GRADIENT_ASCENT',
            },
            inputDataConfig: [{ channelName: 'InputData', dataSource: {
                        s3DataSource: {
                            s3Location: lib_1.S3Location.fromBucket(trainingData, 'data/')
                        }
                    } }],
            outputDataConfig: { s3OutputLocation: lib_1.S3Location.fromBucket(trainingData, 'result/') },
            trainingJobName: 'MyTrainingJob',
        })
    })
});
