// aws-cdk/lib/api/deployment-target.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logging_1 = require("../logging");
const credentials_1 = require("./aws-auth/credentials");
const deploy_stack_1 = require("./deploy-stack");
const toolkit_info_1 = require("./toolkit-info");
exports.DEFAULT_TOOLKIT_STACK_NAME = 'CDKToolkit';
/**
 * Default provisioner (applies to CloudFormation).
 */
class CloudFormationDeploymentTarget {
    constructor(props) {
        this.aws = props.aws;
    }
    async readCurrentTemplate(stack) {
        logging_1.debug(`Reading existing template for stack ${stack.displayName}.`);
        const cfn = await this.aws.cloudFormation(stack.environment.account, stack.environment.region, credentials_1.Mode.ForReading);
        return deploy_stack_1.readCurrentTemplate(cfn, stack.stackName);
    }
    async deployStack(options) {
        const toolkitInfo = await toolkit_info_1.loadToolkitInfo(options.stack.environment, this.aws, options.toolkitStackName || exports.DEFAULT_TOOLKIT_STACK_NAME);
        return deploy_stack_1.deployStack({
            stack: options.stack,
            deployName: options.deployName,
            roleArn: options.roleArn,
            notificationArns: options.notificationArns,
            quiet: options.quiet,
            sdk: this.aws,
            reuseAssets: options.reuseAssets,
            toolkitInfo,
            tags: options.tags,
            execute: options.execute,
            force: options.force
        });
    }
}
exports.CloudFormationDeploymentTarget = CloudFormationDeploymentTarget;
