// @aws-cdk/aws-glue/test/test.database.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const core_1 = require("@aws-cdk/core");
const glue = require("../lib");
module.exports = {
    'default database does not create a bucket'(test) {
        const stack = new core_1.Stack();
        new glue.Database(stack, 'Database', {
            databaseName: 'test_database',
        });
        assert_1.expect(stack).toMatch({
            Resources: {
                DatabaseB269D8BB: {
                    Type: 'AWS::Glue::Database',
                    Properties: {
                        CatalogId: {
                            Ref: "AWS::AccountId"
                        },
                        DatabaseInput: {
                            Name: "test_database"
                        }
                    }
                }
            }
        });
        test.done();
    },
    'explicit locationURI'(test) {
        const stack = new core_1.Stack();
        new glue.Database(stack, 'Database', {
            databaseName: 'test_database',
            locationUri: 's3://my-uri/'
        });
        assert_1.expect(stack).toMatch({
            Resources: {
                DatabaseB269D8BB: {
                    Type: 'AWS::Glue::Database',
                    Properties: {
                        CatalogId: {
                            Ref: "AWS::AccountId"
                        },
                        DatabaseInput: {
                            LocationUri: 's3://my-uri/',
                            Name: "test_database"
                        }
                    }
                }
            }
        });
        test.done();
    },
    'fromDatabase'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        const database = glue.Database.fromDatabaseArn(stack, 'import', 'arn:aws:glue:us-east-1:123456789012:database/db1');
        // THEN
        test.deepEqual(database.databaseArn, 'arn:aws:glue:us-east-1:123456789012:database/db1');
        test.deepEqual(database.databaseName, 'db1');
        test.deepEqual(stack.resolve(database.catalogArn), { 'Fn::Join': ['',
                ['arn:', { Ref: 'AWS::Partition' }, ':glue:', { Ref: 'AWS::Region' }, ':', { Ref: 'AWS::AccountId' }, ':catalog']] });
        test.deepEqual(stack.resolve(database.catalogId), { Ref: 'AWS::AccountId' });
        test.done();
    },
    'locationUri length must be >= 1'(test) {
        const stack = new core_1.Stack();
        test.throws(() => new glue.Database(stack, 'Database', {
            databaseName: 'test_database',
            locationUri: ''
        }));
        test.done();
    },
    'locationUri length must be <= 1024'(test) {
        const stack = new core_1.Stack();
        test.throws(() => new glue.Database(stack, 'Database', {
            databaseName: 'test_database',
            locationUri: 'a'.repeat(1025)
        }));
        test.done();
    }
};
