// @aws-cdk/aws-events-targets/test/aws-api/aws-api.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
const events = require("@aws-cdk/aws-events");
const iam = require("@aws-cdk/aws-iam");
const core_1 = require("@aws-cdk/core");
const targets = require("../../lib");
test('use AwsApi as an event rule target', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const rule = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.expression('rate(15 minutes)')
    });
    // WHEN
    rule.addTarget(new targets.AwsApi({
        service: 'ECS',
        action: 'updateService',
        parameters: {
            service: 'cool-service',
            forceNewDeployment: true
        },
        catchErrorPattern: 'error',
        apiVersion: '2019-01-01',
    }));
    rule.addTarget(new targets.AwsApi({
        service: 'RDS',
        action: 'createDBSnapshot',
        parameters: {
            DBInstanceIdentifier: 'cool-instance'
        },
    }));
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::Events::Rule', {
        Targets: [
            {
                Arn: {
                    "Fn::GetAtt": [
                        "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620",
                        "Arn"
                    ]
                },
                Id: "Target0",
                Input: JSON.stringify({
                    service: 'ECS',
                    action: 'updateService',
                    parameters: {
                        service: 'cool-service',
                        forceNewDeployment: true
                    },
                    catchErrorPattern: 'error',
                    apiVersion: '2019-01-01',
                })
            },
            {
                Arn: {
                    "Fn::GetAtt": [
                        "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620",
                        "Arn"
                    ]
                },
                Id: "Target1",
                Input: JSON.stringify({
                    service: 'RDS',
                    action: 'createDBSnapshot',
                    parameters: {
                        DBInstanceIdentifier: 'cool-instance'
                    },
                })
            }
        ]
    }));
    // Uses a singleton function
    assert_1.expect(stack).to(assert_1.countResources('AWS::Lambda::Function', 1));
    assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Policy', {
        PolicyDocument: {
            Statement: [
                {
                    Action: "ecs:UpdateService",
                    Effect: "Allow",
                    Resource: "*"
                },
                {
                    Action: "rds:CreateDBSnapshot",
                    Effect: "Allow",
                    Resource: "*"
                }
            ],
            Version: "2012-10-17"
        }
    }));
});
test('with policy statement', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const rule = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.expression('rate(15 minutes)')
    });
    // WHEN
    rule.addTarget(new targets.AwsApi({
        service: 'service',
        action: 'action',
        policyStatement: new iam.PolicyStatement({
            actions: ['s3:GetObject'],
            resources: ['resource'],
        })
    }));
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::Events::Rule', {
        Targets: [
            {
                Arn: {
                    "Fn::GetAtt": [
                        "AWSb4cf1abd4e4f4bc699441af7ccd9ec371511E620",
                        "Arn"
                    ]
                },
                Id: "Target0",
                Input: JSON.stringify({
                    service: 'service',
                    action: 'action',
                })
            },
        ]
    }));
    assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Policy', {
        PolicyDocument: {
            Statement: [
                {
                    Action: "s3:GetObject",
                    Effect: "Allow",
                    Resource: "resource"
                },
            ],
            Version: "2012-10-17"
        }
    }));
});
