// @aws-cdk/aws-route53-targets/test/interface-vpc-endpoint-target.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const ec2 = require("@aws-cdk/aws-ec2");
const route53 = require("@aws-cdk/aws-route53");
const core_1 = require("@aws-cdk/core");
const targets = require("../lib");
test('use InterfaceVpcEndpoint as record target', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const vpc = new ec2.Vpc(stack, 'VPC');
    const interfaceVpcEndpoint = new ec2.InterfaceVpcEndpoint(stack, 'InterfaceEndpoint', {
        vpc,
        service: {
            name: 'com.amazonaws.us-west-2.workspaces',
            port: 80
        }
    });
    const zone = new route53.PrivateHostedZone(stack, 'PrivateZone', {
        vpc,
        zoneName: 'test.aws.cdk.com'
    });
    // WHEN
    new route53.ARecord(stack, "AliasEndpointRecord", {
        zone,
        recordName: 'foo',
        target: route53.RecordTarget.fromAlias(new targets.InterfaceVpcEndpointTarget(interfaceVpcEndpoint))
    });
    // THEN
    expect(stack).toHaveResource('AWS::Route53::RecordSet', {
        AliasTarget: {
            HostedZoneId: {
                "Fn::Select": [
                    0,
                    {
                        "Fn::Split": [
                            ":",
                            {
                                "Fn::Select": [
                                    0,
                                    {
                                        "Fn::GetAtt": [
                                            "InterfaceEndpoint12DE6E71",
                                            "DnsEntries"
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                ]
            },
            DNSName: {
                "Fn::Select": [
                    1,
                    {
                        "Fn::Split": [
                            ":",
                            {
                                "Fn::Select": [
                                    0,
                                    {
                                        "Fn::GetAtt": [
                                            "InterfaceEndpoint12DE6E71",
                                            "DnsEntries"
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                ]
            }
        }
    });
});
