// @aws-cdk/aws-events-targets/test/codepipeline/integ.pipeline-event-target.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const codecommit = require("@aws-cdk/aws-codecommit");
const codepipeline = require("@aws-cdk/aws-codepipeline");
const events = require("@aws-cdk/aws-events");
const cdk = require("@aws-cdk/core");
const targets = require("../../lib");
class MockAction {
    constructor(props) {
        this.actionProperties = props;
        this.configuration = props.configuration;
    }
    bind(_scope, _stage, _options) {
        return {
            configuration: this.configuration,
        };
    }
    onStateChange(_name, _target, _options) {
        throw new Error('onStateChange() is not available on MockAction');
    }
}
const app = new cdk.App();
const stack = new cdk.Stack(app, 'pipeline-events');
const repo = new codecommit.Repository(stack, 'Repo', {
    repositoryName: 'TestRepository',
});
const pipeline = new codepipeline.Pipeline(stack, 'pipelinePipeline22F2A91D');
const srcArtifact = new codepipeline.Artifact('Src');
pipeline.addStage({
    stageName: 'Source',
    actions: [new MockAction({
            actionName: 'CodeCommit',
            category: codepipeline.ActionCategory.SOURCE,
            provider: 'CodeCommit',
            artifactBounds: { minInputs: 0, maxInputs: 0, minOutputs: 1, maxOutputs: 1, },
            configuration: {
                RepositoryName: repo.repositoryName,
                BranchName: 'master',
            },
            outputs: [srcArtifact]
        })]
});
pipeline.addStage({
    stageName: 'Build',
    actions: [new MockAction({
            actionName: 'Hello',
            category: codepipeline.ActionCategory.APPROVAL,
            provider: 'Manual',
            artifactBounds: { minInputs: 0, maxInputs: 0, minOutputs: 0, maxOutputs: 0, }
        })]
});
new events.Rule(stack, 'rule', {
    schedule: events.Schedule.expression('rate(1 minute)'),
    targets: [new targets.CodePipeline(pipeline)]
});
app.synth();
