// aws-cdk/lib/api/bootstrap/bootstrap-environment2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cxapi = require("@aws-cdk/cx-api");
const fs = require("fs-extra");
const os = require("os");
const path = require("path");
const __1 = require("..");
async function bootstrapEnvironment2(environment, sdk, toolkitStackName, roleArn, props = {}) {
    var _a, _b, _c, _d;
    if (((_a = props.trustedAccounts) === null || _a === void 0 ? void 0 : _a.length) && !((_b = props.cloudFormationExecutionPolicies) === null || _b === void 0 ? void 0 : _b.length)) {
        throw new Error('--cloudformation-execution-policies are required if --trust has been passed!');
    }
    const outdir = await fs.mkdtemp(path.join(os.tmpdir(), 'cdk-bootstrap-new'));
    const builder = new cxapi.CloudAssemblyBuilder(outdir);
    const templateFile = `${toolkitStackName}.template.json`;
    await fs.copy(path.join(__dirname, 'bootstrap-template.json'), path.join(builder.outdir, templateFile));
    builder.addArtifact(toolkitStackName, {
        type: cxapi.ArtifactType.AWS_CLOUDFORMATION_STACK,
        environment: cxapi.EnvironmentUtils.format(environment.account, environment.region),
        properties: {
            templateFile,
        },
    });
    const assembly = builder.buildAssembly();
    return await __1.deployStack({
        stack: assembly.getStackByName(toolkitStackName),
        sdk,
        roleArn,
        tags: props.tags,
        execute: props.execute,
        parameters: {
            FileAssetsBucketName: props.bucketName,
            FileAssetsBucketKmsKeyId: props.kmsKeyId,
            TrustedAccounts: (_c = props.trustedAccounts) === null || _c === void 0 ? void 0 : _c.join(','),
            CloudFormationExecutionPolicies: (_d = props.cloudFormationExecutionPolicies) === null || _d === void 0 ? void 0 : _d.join(','),
        },
    });
}
exports.bootstrapEnvironment2 = bootstrapEnvironment2;
