// cdk8s/lib/api-object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@aws-cdk/core");
const chart_1 = require("./chart");
class ApiObject extends core_1.Construct {
    /**
     * Defines an API object.
     *
     * @param scope the construct scope
     * @param ns namespace
     * @param options options
     */
    constructor(scope, ns, options) {
        var _a, _b;
        super(scope, ns);
        this.options = options;
        this.chart = chart_1.Chart.of(this);
        this.kind = options.kind;
        this.apiVersion = options.apiVersion;
        this.name = (_b = (_a = options.metadata) === null || _a === void 0 ? void 0 : _a.name, (_b !== null && _b !== void 0 ? _b : this.chart.generateObjectName(this)));
    }
    /**
     * Renders the object to Kubernetes config.
     * @internal
     */
    _render() {
        return {
            ...this.options,
            metadata: {
                ...this.options.metadata,
                name: this.name
            }
        };
    }
}
exports.ApiObject = ApiObject;
