// cdk8s/lib/chart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@aws-cdk/core");
const fs = require("fs");
const path = require("path");
const api_object_1 = require("./api-object");
const YAML = require("yaml");
const _tokens_1 = require("./_tokens");
const _util_1 = require("./_util");
const names_1 = require("./names");
class Chart extends core_1.Construct {
    constructor(scope, ns) {
        super(scope, ns);
        this.manifestFile = `${this.node.uniqueId}.k8s.yaml`;
    }
    /**
     * Finds the chart in which a node is defined.
     * @param node a construct node
     */
    static of(node) {
        if (node instanceof Chart) {
            return node;
        }
        const parent = node.node.scope;
        if (!parent) {
            throw new Error(`cannot find a parent chart (directly or indirectly)`);
        }
        return Chart.of(parent);
    }
    /**
     * Generates a app-unique name for an object given it's construct node path.
     *
     * Different resource types may have different constraints on names
     * (`metadata.name`). The previous version of the name generator was
     * compatible with DNS_SUBDOMAIN but not with DNS_LABEL.
     *
     * For example, `Deployment` names must comply with DNS_SUBDOMAIN while
     * `Service` names must comply with DNS_LABEL.
     *
     * Since there is no formal specification for this, the default name
     * generation scheme for kubernetes objects in cdk8s was changed to DNS_LABEL,
     * since it’s the common denominator for all kubernetes resources
     * (supposedly).
     *
     * You can override this method if you wish to customize object names at the
     * chart level.
     *
     * @param apiObject The API object to generate a name for.
     */
    generateObjectName(apiObject) {
        return names_1.Names.toDnsLabel(apiObject.node.path);
    }
    synthesize(session) {
        const resources = new Array();
        for (const resource of this.node.findAll()) {
            if (!(resource instanceof api_object_1.ApiObject)) {
                continue;
            }
            const manifest = _util_1.removeEmpty(_tokens_1.resolve(this, resource._render()));
            resources.push(manifest);
        }
        // convert each resource to yaml and separate with a '---' line
        const doc = resources.map(r => toYaml(r)).join('---\n');
        fs.writeFileSync(path.join(session.assembly.outdir, this.manifestFile), doc);
    }
}
exports.Chart = Chart;
function toYaml(o) {
    // lose anchors which are based on reference equality
    const x = JSON.parse(JSON.stringify(o));
    return YAML.stringify(x);
}
