// @aws-cdk/aws-ecs/test/ec2/test.cross-stack.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const ec2 = require("@aws-cdk/aws-ec2");
const elbv2 = require("@aws-cdk/aws-elasticloadbalancingv2");
const core_1 = require("@aws-cdk/core");
const ecs = require("../../lib");
// Test various cross-stack Cluster/Service/ALB scenario's
let app;
let stack1;
let stack2;
let cluster;
let service;
function expectIngress(stack) {
    assert_1.expect(stack).to(assert_1.haveResource('AWS::EC2::SecurityGroupIngress', {
        FromPort: 32768,
        ToPort: 65535,
        GroupId: { "Fn::ImportValue": "Stack1:ExportsOutputFnGetAttClusterDefaultAutoScalingGroupInstanceSecurityGroup1D15236AGroupIdEAB9C5E1" },
    }));
}
module.exports = {
    "setUp"(cb) {
        app = new core_1.App();
        stack1 = new core_1.Stack(app, 'Stack1');
        const vpc = new ec2.Vpc(stack1, 'Vpc');
        cluster = new ecs.Cluster(stack1, 'Cluster', {
            vpc,
            capacity: { instanceType: new ec2.InstanceType('t2.micro'), }
        });
        stack2 = new core_1.Stack(app, 'Stack2');
        const taskDefinition = new ecs.Ec2TaskDefinition(stack2, 'TD');
        const container = taskDefinition.addContainer('Main', {
            image: ecs.ContainerImage.fromRegistry('asdf'),
            memoryLimitMiB: 512
        });
        container.addPortMappings({ containerPort: 8000 });
        service = new ecs.Ec2Service(stack2, 'Service', {
            cluster,
            taskDefinition,
        });
        cb();
    },
    "ALB next to Service"(test) {
        // WHEN
        const lb = new elbv2.ApplicationLoadBalancer(stack2, "ALB", { vpc: cluster.vpc });
        const listener = lb.addListener("listener", { port: 80 });
        listener.addTargets("target", {
            port: 80,
            targets: [service]
        });
        // THEN: it shouldn't throw due to cyclic dependencies
        assert_1.expect(stack2).to(assert_1.haveResource('AWS::ECS::Service'));
        expectIngress(stack2);
        test.done();
    },
    "ALB next to Cluster"(test) {
        // WHEN
        const lb = new elbv2.ApplicationLoadBalancer(stack1, "ALB", { vpc: cluster.vpc });
        const listener = lb.addListener("listener", { port: 80 });
        listener.addTargets("target", {
            port: 80,
            targets: [service]
        });
        // THEN: it shouldn't throw due to cyclic dependencies
        assert_1.expect(stack2).to(assert_1.haveResource('AWS::ECS::Service'));
        expectIngress(stack2);
        test.done();
    },
    "ALB in its own stack"(test) {
        // WHEN
        const stack3 = new core_1.Stack(app, 'Stack3');
        const lb = new elbv2.ApplicationLoadBalancer(stack3, "ALB", { vpc: cluster.vpc });
        const listener = lb.addListener("listener", { port: 80 });
        listener.addTargets("target", {
            port: 80,
            targets: [service]
        });
        // THEN: it shouldn't throw due to cyclic dependencies
        assert_1.expect(stack2).to(assert_1.haveResource('AWS::ECS::Service'));
        expectIngress(stack2);
        test.done();
    },
};
